File: leetcode/medium/576_out_of_boundary_paths.py
# -----------------------------------------
# My Solution: Dynamic Programming
#
# Time Complexity: O(mn * maxMove)
# -----------------------------------------

# -----> Version 1: Space Complexity: O(mn * maxMove)
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        MOD = 10 ** 9 + 7

        paths_move_out = [[0] * n for _ in range(m)]
        for row in range(m):
            for col in [0, n - 1]:
                paths_move_out[row][col] += 1
        for row in [0, m - 1]:
            for col in range(n):
                paths_move_out[row][col] += 1

        # dp[i][j][k] := total paths to move to outside boundary starts from (i, j) with maxMove k
        dp = [[[0] * (maxMove + 1) for _ in range(n)] for _ in range(m)]
        for move in range(1, maxMove + 1):
            for row in range(m):
                for col in range(n):
                    dp[row][col][move] += paths_move_out[row][col]
                    for row_move, col_move in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                        prev_row = row + row_move
                        prev_col = col + col_move
                        if 0 <= prev_row < m and 0 <= prev_col < n:
                            dp[row][col][move] = (dp[row][col][move] + dp[prev_row][prev_col][move - 1]) % MOD

        return dp[startRow][startColumn][-1]

# -----> Version 2: Space Complexity: O(mn)
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        MOD = 10 ** 9 + 7

        paths_move_out = [[0] * n for _ in range(m)]
        for row in range(m):
            for col in [0, n - 1]:
                paths_move_out[row][col] += 1
        for row in [0, m - 1]:
            for col in range(n):
                paths_move_out[row][col] += 1

        curr_dp = [[0] * n for _ in range(m)]
        for _ in range(maxMove):
            next_dp = [[paths_move_out[row][col] for col in range(n)] for row in range(m)]
            for row in range(m):
                for col in range(n):
                    for row_move, col_move in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                        prev_row = row + row_move
                        prev_col = col + col_move
                        if 0 <= prev_row < m and 0 <= prev_col < n:
                            next_dp[row][col] = (next_dp[row][col] + curr_dp[prev_row][prev_col]) % MOD
            curr_dp = next_dp

        return curr_dp[startRow][startColumn]

File: leetcode/medium/576_out_of_boundary_paths.cpp
// -----------------------------------------
// My Solution
//
// Time  Complexity: O(mnr^2)
// Space Complexity: O(mnr)
// -----------------------------------------
// r := maxMove
class Solution {
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        row_size = m;
        col_size = n;
        count_paths_cache = vector<vector<vector<int>>>(m + 1, vector<vector<int>>(n + 1, vector<int>(maxMove + 1, -1)));

        long long total_paths = 0;
        for (int remain_moves = 1; remain_moves <= maxMove; ++remain_moves)
            total_paths = (total_paths + count_paths(startRow, startColumn, remain_moves)) % MOD;
        return total_paths;
    }

private:
    const int MOD = 1e9 + 7;
    const int position_moves[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

    int row_size;
    int col_size;
    vector<vector<vector<int>>> count_paths_cache;

    bool position_valid(int row, int col) {
        return 0 <= row && row < row_size && 0 <= col && col < col_size;
    }

    long long count_paths(int row, int col, int remain_moves) {
        if (remain_moves == 0)
            return 0;
        if (count_paths_cache[row][col][remain_moves] != -1)
            return count_paths_cache[row][col][remain_moves];

        long long paths = 0;
        for (auto& move : position_moves) {
            int next_row = row + move[0];
            int next_col = col + move[1];
            if (remain_moves == 1 && !position_valid(next_row, next_col)) {
                ++paths;
            } else if (remain_moves > 1 && position_valid(next_row, next_col)) {
                paths = (paths + count_paths(next_row, next_col, remain_moves - 1)) % MOD;
            }
        }

        count_paths_cache[row][col][remain_moves] = paths;
        return paths;
    }
};

// -----------------------------------------
// My Solution: Dynamic Programming
//
// Time  Complexity: O(mnr)
// Space Complexity: O(mn)
// -----------------------------------------
// r := maxMove
class Solution {
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        const int MOD = 1e9 + 7;
        const int position_moves[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

        vector<vector<int>> curr_dp(m, vector<int>(n, 0));
        for (int moves = 1; moves <= maxMove; ++moves) {
            vector<vector<int>> next_dp(m, vector<int>(n, 0));
            for (int row = 0; row < m; ++row) {
                for (int col = 0; col < n; ++col) {
                    if (row == 0)
                        ++next_dp[row][col];
                    if (col == 0)
                        ++next_dp[row][col];
                    if (row == m - 1)
                        ++next_dp[row][col];
                    if (col == n - 1)
                        ++next_dp[row][col];

                    for (auto& move : position_moves) {
                        int prev_row = row + move[0];
                        int prev_col = col + move[1];
                        if (0 <= prev_row && prev_row < m && 0 <= prev_col && prev_col < n)
                            next_dp[row][col] = (next_dp[row][col] + curr_dp[prev_row][prev_col]) % MOD;
                    }
                }
            }

            curr_dp = next_dp;
        }
        return curr_dp[startRow][startColumn];
    }
};
